app/src/main/java/org/p2p/wallet/striga/offramp/ui/StrigaOffRampPresenter.kt

```diff
@@ -0,0 +1,344 @@
+            viewState.collect {
+                view?.setRatioState(rateTickerMapper.mapRate(it.exchangeRateState))
+                view?.setButtonState(it.buttonState)
+            }
+        }
+    }
+
+    override fun onTokenAAmountChange(amountA: String) {
+        inputAmountA = amountA.toBigDecimalOrZero()
+        setAndCalcTokenAmount(
+            targetTokenType = StrigaOffRampTokenType.TokenB,
+            amount = inputAmountA
+        )
+        validate()
+    }
+
+    override fun onTokenBAmountChange(amountB: String) {
+        inputAmountB = amountB.toBigDecimalOrZero()
+        setAndCalcTokenAmount(
+            targetTokenType = StrigaOffRampTokenType.TokenA,
+            amount = inputAmountB
+        )
+        validate()
+    }
+
+    override fun onAllAmountClick() {
+        // ignore all amount click when no internet or rate is not loaded/error occurred
+        if (!connectionManager.connectionStatus.value || cannotClickAllAmount) return
+
+        setTokenAmount(
+            targetTokenType = StrigaOffRampTokenType.TokenA,
+            amount = balance
+        )
+        setAndCalcTokenAmount(
+            targetTokenType = StrigaOffRampTokenType.TokenB,
+            amount = balance
+        )
+        validate()
+    }
+
+    override fun onSubmit() {
+        // todo: https://p2pvalidator.atlassian.net/browse/PWN-9267
+    }
+
+    private fun setTokenAmount(targetTokenType: StrigaOffRampTokenType, amount: BigDecimal) {
+        when (targetTokenType) {
+            StrigaOffRampTokenType.TokenA -> {
+                inputAmountA = amount
+                val tokenAState = swapWidgetMapper.mapByState(
+                    tokenType = StrigaOffRampTokenType.TokenA,
+                    state = swapWidgetMapper.mapTokenA(
+                        amount = amount,
+                        amountAvailable = balance,
+                    )
+                )
+                view?.setTokenAWidgetState(tokenAState)
+            }
+            StrigaOffRampTokenType.TokenB -> {
+                inputAmountB = amount
+                val tokenBState = swapWidgetMapper.mapTokenB(amount)
+                view?.setTokenBWidgetState(
+                    swapWidgetMapper.mapByState(
+                        tokenType = StrigaOffRampTokenType.TokenB,
+                        state = tokenBState
+                    )
+                )
+            }
+        }
+    }
+
+    private fun setAndCalcTokenAmount(targetTokenType: StrigaOffRampTokenType, amount: BigDecimal) {
+        val calculatedAmount = interactor.calculateAmountByRate(targetTokenType, rate, amount)
+        setTokenAmount(targetTokenType, calculatedAmount)
+    }
+
+    private fun validate() {
+        val buttonState = interactor.validateAmount(
+            amountA = inputAmountA,
+            amountB = inputAmountB,
+            amountAvailable = balance
+        )
+        setButtonState(buttonState)
+
+        if (buttonState.isAmountError) {
```

Comment:
the StrigaOffRampButtonState might be renamed to something else, because it is particularly represents the error state